llvm/tools/sycl-post-link/CUDASpecConstantToSymbol.cpp

```diff
@@ -0,0 +1,267 @@
+//===-- CUDASpecConstantToSymbol.cpp - CUDA Spec Constants To Symbol Pass -===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// See comments in the header.
+//===----------------------------------------------------------------------===//
+
+#include "CUDASpecConstantToSymbol.h"
+#include "Support.h"
+#include "llvm/ADT/Statistic.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/SYCLLowerIR/TargetHelpers.h"
+#include "llvm/Support/MathExtras.h"
+
+#ifndef NDEBUG
+#include "llvm/Support/Debug.h"
+#endif
+
+using namespace llvm;
+
+#define DEBUG_TYPE "CUDASpecConstantToSymbol"
+
+STATISTIC(TotalSymbolSize,
+          "amount of global memory allocated for specialization constants.");
+
+static GlobalVariable *createGlobal(Module &M, size_t Size, StringRef Name,
+                                    bool RoundUp) {
+  size_t AllocSize = PowerOf2Ceil(Size);
+  if (RoundUp && AllocSize < 64)
+    // cuMemcpyHtoD has problems copying to symbols that are not power of two
+    // or less than 64-bit in size.
+    AllocSize = 64;
+
+  LLVM_DEBUG(llvm::dbgs() << "\tallocating: " << AllocSize << " bytes\n");
+  TotalSymbolSize += AllocSize;
+
+  auto *ArrTy = ArrayType::get(Type::getInt8Ty(M.getContext()), AllocSize);
+  const bool IsConstant = true;
+  const bool ExternallyInitialized = true;
+  const unsigned AS = 4; // ADDRESS_SPACE_CONST
+  return new GlobalVariable(M, ArrTy, IsConstant, GlobalValue::WeakODRLinkage,
+                            Constant::getNullValue(ArrTy), Name, nullptr,
+                            GlobalValue::NotThreadLocal, AS,
+                            ExternallyInitialized);
+}
+
+static unsigned uintFromMDNode(const MDNode *Node, unsigned OpNumber) {
+  auto *CMD =
+      cast<ConstantAsMetadata>(Node->getOperand(OpNumber).get())->getValue();
+  return static_cast<unsigned>(CMD->getUniqueInteger().getZExtValue());
```

Comment:
```suggestion
  return static_cast<unsigned>(mdconst::extract<ConstantInt>(
       Node->getOperand(OpNumber).get())->getZExtValue());
```

[Doxygen](https://llvm.org/doxygen/namespacellvm_1_1mdconst.html#ad938857d6c6603847adf3a8cbe403d17)

Reply:
Nice.